### ISSUE TITLE
Cannot use per-function docstring in help messages when argparse description is not set

### ISSUE THREAD

**Issue by xNinjaKittyx** [python-cmd2/cmd2#522]
I have been noticing help messages of only a single docstring will apply to its argparser. So any commands that use a common parser will always have the same docstring, which is fairly inconvenient if someone wants to use the same parser for multiple commands.

```Python3
import argparse

from cmd2 import Cmd, with_argparser


parser = argparse.ArgumentParser()
parser.add_argument('-s', '--something', type=str, help='some argument')


class SomeClass(Cmd):
    @with_argparser(parser)
    def do_foo(self, opts):
        """This is foo docstring"""
        pass


    @with_argparser(parser)
    def do_bar(self, opts):
        """This is bar docstring"""
        pass


if __name__ == "__main__":
    SomeClass().cmdloop()
```

Now using the Cmd
```
(Cmd) foo -h
usage: bar [-h] [-s SOMETHING]

This is foo docstring

optional arguments:
  -h, --help            show this help message and exit
  -s SOMETHING, --something SOMETHING
                        some argument
(Cmd) bar -h
usage: bar [-h] [-s SOMETHING]

This is foo docstring

optional arguments:
  -h, --help            show this help message and exit
  -s SOMETHING, --something SOMETHING
                        some argument
(Cmd)
```

~~Looking at the code, it could be fixed by setting the argparser's description inside the function wrapper, rather than outside of it. I don't know how this will affect the rest of the code though.~~
Nevermind, it's not that straightforward. There needs to be a way to know if argparse.description was not declared on init vs declaring it dynamically.

If people want all of their functions to have the same help message, they can always set the parser's description. However, if people want to have unique messages for each one, it would be nice to be setting that docstring based on what function is being used.

**Comment by tleonhardt** [comment]
@anselor @kmvanbrunt Would either of you want to take a look at this?  It sounds up your alley.

**Comment by kmvanbrunt** [comment]
This isn't really a cmd2 or argparse issue. The same parser object is being passed into both function wrappers. Even if you did change the description in between the function declarations, both functions would still print the new description because the same object is being altered.

A deep copy of the parser needs to be made for each function you wish to use it with before the description gets set. I tried this in Python 3.5 on Debian 9 and didn't have much luck.

I received lots of exceptions doing this:
`parser_copy = copy.deepcopy(orig_parser)`

**Comment by kmvanbrunt** [comment]
I found a decent workaround if deepcopy doesn't work. Write a function to create and return a copy of the type of parser you want to reuse. Then call that function for each parser copy that you need.

```Python
def make_parser():
        new_parser = ACArgumentParser()
        new_parser.add_argument('-a', '--all', action='store_true', help='Handle all settings')
        return new_parser
```

**Comment by xNinjaKittyx** [comment]
Sure, that seems like a reasonable workaround.

The way I originally was thinking of was something like this:
```
def with_argparser(argparser: argparse.ArgumentParser) -> Callable:
    """A decorator to alter a cmd2 method to populate its ``args`` argument by parsing arguments
    with the given instance of argparse.ArgumentParser.

    :param argparser: argparse.ArgumentParser - given instance of ArgumentParser
    :return: function that gets passed parsed args
    """
    import functools

    # noinspection PyProtectedMember
    def arg_decorator(func: Callable):
        if argparser.description is None:
            argparser.init_with_description = False
        @functools.wraps(func)
        def cmd_wrapper(instance, cmdline):
            lexed_arglist = parse_quoted_string(cmdline)
            try:
                # argparser defaults the program name to sys.argv[0]
                # we want it to be the name of our command
                argparser.prog = func.__name__[3:]
                # If the description has not been set, then use the method docstring if one exists

                if not argparser.init_with_description and func.__doc__:
                    argparser.description = func.__doc__
                args = argparser.parse_args(lexed_arglist)
            except SystemExit:
                return
            else:
                return func(instance, args)

        if func.__doc__:
            setattr(cmd_wrapper, HELP_SUMMARY, func.__doc__)

        cmd_wrapper.__doc__ = argparser.format_help()

        # Mark this function as having an argparse ArgumentParser
        setattr(cmd_wrapper, 'argparser', argparser)

        return cmd_wrapper

    return arg_decorator
```

Where the function name and doc string were dynamically changed based on the command.
Your suggestion is okay, and it means each command has to have its own unique argparser object, which is totally reasonable.

Feel free to close based on your decision.

**Comment by tleonhardt** [comment]
@kmvanbrunt For whatever its worth, the inability to **deepcopy** an ``argparse.ArgumentParser`` is a known issue and has been fixed in Python 3.7.  Unfortunately, I don't believe there are any plans to back port that fix to Python 3.5 or 3.6.  

I would prefer for ``cmd2`` to handle correct behavior within the argparse-based decorators if it is reasonable.  If not, then we need to add a big bold glaring warning to the documentation telling end users that each command needs a unique ``ArgumentParser``.
